### ISSUE TITLE
JaCoCo fails to build a report too frequently

### ISSUE THREAD

**Issue by wissenstein** [jacoco/jacoco#546]
### Steps to reproduce

JaCoCo version: 0.7.9
Operating system: Windows 10 version 10.0 running on amd64
Tool integration: Maven

I am trying to build a project with 152 classes.

### Expected behaviour

The broject is built successfully.

### Actual behaviour

Some times the project is built successfully, but very frequently it fails with the following exception (may fail 8 times or more before a successful build with the same source):

```
--- jacoco-maven-plugin:0.7.9:report (post-unit-test) @ BusinessTravelSystem ---
Loading execution data file C:\projects\bts\project\business-travel-system\BusinessTravelSystem\target\coverage-reports\jacoco-ut.exec
------------------------------------------------------------------------
BUILD FAILURE
------------------------------------------------------------------------
Total time: 01:02 min
Finished at: 2017-06-01T14:33:39+02:00
Final Memory: 67M/581M
------------------------------------------------------------------------
Failed to execute goal org.jacoco:jacoco-maven-plugin:0.7.9:report (post-unit-test) on project BusinessTravelSystem: An error has occurred in JaCoCo report generation. Error while creating report: null: EOFException -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.jacoco:jacoco-maven-plugin:0.7.9:report (post-unit-test) on project BusinessTravelSystem: An error has occurred in JaCoCo report generation.
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:212)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:116)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:80)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:307)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:193)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:106)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:863)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:288)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:199)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: org.apache.maven.plugin.MojoExecutionException: An error has occurred in JaCoCo report generation.
	at org.jacoco.maven.AbstractReportMojo.execute(AbstractReportMojo.java:167)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:207)
	... 20 more
Caused by: org.apache.maven.reporting.MavenReportException: Error while creating report: null
	at org.jacoco.maven.AbstractReportMojo.executeReport(AbstractReportMojo.java:183)
	at org.jacoco.maven.AbstractReportMojo.execute(AbstractReportMojo.java:165)
	... 22 more
Caused by: java.io.EOFException
	at java.io.DataInputStream.readByte(DataInputStream.java:267)
	at org.jacoco.core.internal.data.CompactDataInput.readBooleanArray(CompactDataInput.java:63)
	at org.jacoco.core.data.ExecutionDataReader.readExecutionData(ExecutionDataReader.java:149)
	at org.jacoco.core.data.ExecutionDataReader.readBlock(ExecutionDataReader.java:115)
	at org.jacoco.core.data.ExecutionDataReader.read(ExecutionDataReader.java:92)
	at org.jacoco.core.tools.ExecFileLoader.load(ExecFileLoader.java:59)
	at org.jacoco.core.tools.ExecFileLoader.load(ExecFileLoader.java:73)
	at org.jacoco.maven.ReportSupport.loadExecutionData(ReportSupport.java:87)
	at org.jacoco.maven.ReportMojo.loadExecutionData(ReportMojo.java:60)
	at org.jacoco.maven.AbstractReportMojo.executeReport(AbstractReportMojo.java:177)
	... 23 more
```

**Comment by Godin** [comment]
@wissenstein `EOFException` indicates that exec file is truncated, i.e. wasn't fully written by an agent. Agent writes exec file at JVM shutdown using `Runtime.addShutdownHook()` API. So please make sure that JVM with tests terminates gracefully and shutdown hooks are executed.

Should also be noted that due to race condition in JDKs up to 9 EA b125 shutdown hooks not always executed if main thread is left with an interrupted status - see https://github.com/jacoco/jacoco/issues/394#issuecomment-208531845 So please also make sure that your tests do not leave main thread with an interrupted status.

**Comment by JyothiAswathaiah** [comment]
Hi, 
We are also facing exactly the same issue. It happens if there are more number of tests and it happens intermittently. Is there any work around for this?

**Comment by Godin** [comment]
@JyothiAswathaiah sorry, but workaround for what? This ticket does not describe root cause and initial comment https://github.com/jacoco/jacoco/issues/546#issue-232873722 as well as yours does not provide any information to understand the root cause. So either provide us more information (ideally reproducer) to be able to understand your issue. Or as was said in comment https://github.com/jacoco/jacoco/issues/546#issuecomment-305495921 :

> EOFException indicates that exec file is truncated, i.e. wasn't fully written by an agent. Agent writes exec file at JVM shutdown using Runtime.addShutdownHook() API. So **please make sure that JVM with tests terminates gracefully** and shutdown hooks are executed.

in other words - that nothing in your setup leads to truncation/corruption of exec file.

**Comment by Godin** [comment]
As was described earlier there is not enough information to understand the reason of truncated file, so I'm closing this ticket. But feel free to comment, if you can provide such information.

**Comment by JyothiAswathaiah** [comment]
Thanks for the information. The thread was indeed in interrupted state, we have fixed the cause for the thread to be in interrupted state. Jacoco is now generating the right coverage always! :)
